sync/class.jetpack-sync-module-import.php

```diff
@@ -2,86 +2,153 @@
 	public function init_listeners( $callable ) {
 		add_action( 'export_wp', $callable );
+		add_action( 'jetpack_sync_import_start', $callable, 10, 2 );
 		add_action( 'jetpack_sync_import_end', $callable, 10, 2 );
 
-		// Movable type, RSS, Livejournal
-		add_action( 'import_done', array( $this, 'sync_import_done' ) );
+		// WordPress.
+		add_action( 'import_start', array( $this, 'sync_import_action' ) );
 
-		// WordPress, Blogger, Livejournal, woo tax rate
-		add_action( 'import_end', array( $this, 'sync_import_end' ) );
+		// Movable type, RSS, Livejournal.
+		add_action( 'import_done', array( $this, 'sync_import_action' ) );
+
+		// WordPress, Blogger, Livejournal, woo tax rate.
+		add_action( 'import_end', array( $this, 'sync_import_action' ) );
 	}
 
 	public function set_defaults() {
-		$this->import_end = false;
+		$this->synced_actions = array();
 	}
 
-	public function sync_import_done( $importer ) {
-		// We already ran an send the import
-		if ( $this->import_end ) {
-			return;
-		}
-
-		$importer_name = $this->get_importer_name( $importer );
-
-		/**
-		 * Sync Event that tells that the import is finished
-		 *
-		 * @since 5.0.0
-		 *
-		 * $param string $importer
-		 */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
-		$this->import_end = true;
-	}
+	public function sync_import_action( $importer ) {
+		$import_action = current_filter();
+		// Map action to event name.
+		$sync_action = self::$import_sync_action_map[ $import_action ];
 
-	public function sync_import_end() {
-		// We already ran an send the import
-		if ( $this->import_end ) {
+		// Only sync each action once per import.
+		if ( array_key_exists( $sync_action, $this->synced_actions ) && $this->synced_actions[ $sync_action ] ) {
 			return;
 		}
 
-		$this->import_end = true;
-		$importer         = 'unknown';
-		$backtrace        = wp_debug_backtrace_summary( null, 0, false );
-		if ( $this->is_importer( $backtrace, 'Blogger_Importer' ) ) {
-			$importer = 'blogger';
-		}
+		// Mark this action as synced.
+		$this->synced_actions[ $sync_action ] = true;
 
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WC_Tax_Rate_Importer' ) ) {
-			$importer = 'woo-tax-rate';
+		// Prefer self-reported $importer value.
+		if ( ! $importer ) {
+			// Fall back to inferring by calling class name.
+			$importer = self::get_calling_importer_class();
 		}
 
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WP_Import' ) ) {
-			// phpcs:ignore WordPress.WP.CapitalPDangit
-			$importer = 'wordpress';
+		// Get $importer from known_importers.
+		if ( isset( self::$known_importers[ $importer ] ) ) {
+			$importer = self::$known_importers[ $importer ];
 		}
 
 		$importer_name = $this->get_importer_name( $importer );
 
-		/** This filter is already documented in sync/class.jetpack-sync-module-posts.php */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
+		do_action( $sync_action, $importer, $importer_name );
```

Comment:
It would be really great to leave a comment here. So when someone searched for `jetpack_sync_import_start` or `jetpack_sync_import_end` we also get this line.

Reply:
Great suggestion. Comment added!